[flang] Modifications to ieee floating point environment procedures

vdonaldson opened the pull request:
Intrinsic module procedures ieee_get_modes, ieee_set_modes, ieee_get_status, and ieee_set_status store and retrieve opaque data values whose size varies by machine and OS environment. These data values are usually, but not always small. Their sizes are not directly known in a cross compilation environment. Address this issue by implementing two mechanisms for processing these data values. Environments that use typical small data sizes can access storage defined at compile time. When this is not valid, data storage of any size can be allocated at runtime.

rorth commented:
This doesn't compile on Solaris:
```
/vol/llvm/src/llvm-project/local/flang/runtime/exceptions.cpp:114:5: error: function-like macro '__GLIBC_USE' is not defined
  114 | #if __GLIBC_USE(IEC_60559_BFP_EXT)
      |     ^
```

vdonaldson commented:
> This doesn't compile on Solaris:
> 
> ```
> /vol/llvm/src/llvm-project/local/flang/runtime/exceptions.cpp:114:5: error: function-like macro '__GLIBC_USE' is not defined
>   114 | #if __GLIBC_USE(IEC_60559_BFP_EXT)
>       |     ^
> ```

Thanks for checking that. Is `femode_t` defined in `/usr/include/bits/fenv.h`? If yes, is it under some preprocessor directive?

kkwli commented:
> This doesn't compile on Solaris:
> 
> ```
> /vol/llvm/src/llvm-project/local/flang/runtime/exceptions.cpp:114:5: error: function-like macro '__GLIBC_USE' is not defined
>   114 | #if __GLIBC_USE(IEC_60559_BFP_EXT)
>       |     ^
> ```

I got the same error on AIX. And `/usr/include/fenv.h` on AIX does not define `femode_t` either.

kkwli commented:
> > > This doesn't compile on Solaris:
> > > ```
> > > /vol/llvm/src/llvm-project/local/flang/runtime/exceptions.cpp:114:5: error: function-like macro '__GLIBC_USE' is not defined
> > >   114 | #if __GLIBC_USE(IEC_60559_BFP_EXT)
> > >       |     ^
> > > ```
> > 
> > 
> > I got the same error on AIX. And `/usr/include/fenv.h` on AIX does not define `femode_t` either.
> 
> @rorth, @kkwli - Could you please check if this iteration of the code builds?

It builds successfully on AIX. Thanks.

jeanPerier reviewed (approved):
Please wait for the feedback from @rorth on your latest iteration. Lowering code looks good to me.

rorth commented:
While this version builds on Solaris (both sparcv9 and amd64) and shows clean test results on Solaris/amd64, SPARC is different: on both Solaris/sparcv9 (which lacks `femode_t`) and Linux/sparc64 (which has it) I get two test failures:
```
  Flang :: Lower/Intrinsics/ieee_femodes.f90
  Flang :: Lower/Intrinsics/ieee_festatus.f90
```
On both Solaris and LInux I get
```
/vol/llvm/src/llvm-project/local/flang/test/Lower/Intrinsics/ieee_femodes.f90:35:11: error: CHECK: expected string not found in input
 ! CHECK: %[[V_98:[0-9]+]] = fir.convert %[[V_60]]#1 : (!fir.ref<!fir.type<_QM__fortran_ieee_exceptionsTieee_modes_type{_QM__fortran_ieee_exceptionsTieee_modes_type.__data:!fir.array<2xi32>,_QM__fortran_ieee_exceptionsTieee_modes_type.__allocatable_data:!fir.box<!fir.heap<!fir.array<?xi8>>>}>>) -> !fir.ref<i32>
          ^
<stdin>:85:37: note: scanning from here
 fir.store %77 to %75 : !fir.ref<i8>
                                    ^
<stdin>:85:37: note: with "V_60" equal to "60" 
 fir.store %77 to %75 : !fir.ref<i8>
                                    ^
```

rorth commented:
I don't think the last build fix is right:
- The two tests `PASS` on Solaris/amd64, which is also`system-solaris`.
- Likewise, the tests `FAIL` on Linux/sparc64, although that does have `femode_t`.

vdonaldson commented:
@kkwli - Thanks for the checking the AIX build.

@rorth - Thanks for checking solaris builds. The generated code should be different for these tests for at least some sparc environments. I've updated the test with a different UNSUPPORTED directive to address that. Could you please check if that eliminates the test failure? If not, we need to find a directive that is applicable for this scenario.

rorth commented:
As I'd already mentioned, `system-solaris` is wrong.  Besides, if you ever consider using sparc-specific features, `*-target-arch` won't work: it only exists in `compiler-rt/test`, certainly not in `flang/test`.

vdonaldson commented:
> I don't think the last build fix is right:
> 
>     * The two tests `PASS` on Solaris/amd64, which is also`system-solaris`.
> 
>     * Likewise, the tests `FAIL` on Linux/sparc64, although that does have `femode_t`.

This PR allows different code to be generated for sparc / solaris environments vs. other environments since the relevant data sizes are substantially different. I expect the code should be correct for the various cases, but I have no way to verify that myself. Given that, for the tests I would like to find a directive that suppresses the tests for environments that generate alternate code.

vdonaldson commented:
> > This PR allows different code to be generated for sparc / solaris environments vs. other environments since the relevant data sizes are substantially different. I expect the code should be correct for the various cases, but I have no way to verify that myself. Given that, for the tests I would like to find a directive that suppresses the tests for environments that generate alternate code.
> 
> Actually, you can verify things on SPARC yourself: there's both a Solaris/sparcv9 and a Linux/sparc64 system in the [Compile Farm](https://portal.cfarm.net) (and many other systems on top of those).
> 
> Besides, if you can tell me **exactly** what to check, I could do this for both Solaris and Linux. I think that would be better than blindly `XFAIL`ing the tests without understanding the issues.

The two tests are executable Fortran tests. Output for both tests lists expected output on the left side of each line, and actual output on the right. Is that output correct?

vdonaldson commented:
Thanks @rorth and @kkwli for help with resolving build issues.

The intent of this PR is to allow more flexibility in how these four floating point environment procedures are implemented, and use that flexibility to solve several problems. I believe it now does that without breaking builds or introducing regressions to working functionality in any environment. I am less concerned with lit test coverage, although the included tests address common use cases. I propose to push this iteration of the code. Anyone may extend or improve tests or code as occasion arises.

kkwli reviewed (approved):
LGTM

rorth commented:
I've now retested the latest version on 4 targets, but it's still not right:
- While the patch compiles and all tests `PASS` on both `x86_64-pc-linux-gnu` and `amd64-pc-solaris2.11`
- on both `sparcv9-sun-solaris2.11` and `sparc64-unknown-linux-gnu`, I get the same failures as before:
  ```
  Flang :: Lower/Intrinsics/ieee_femodes.f90
  Flang :: Lower/Intrinsics/ieee_festatus.f90
  ```

This happens because the `REQUIRES: x86-registered-target` doesn't do what you think it does: it checks if `flang` is configured to support the X86 target, which is true in both my targets (configured with `-DLLVM_TARGETS_TO_BUILD=all`).  However, you make no use whatsoever of this information: `bbc` is still invoked without an explicit `-target` option, so runs for the native/default target instead, which is `sparc*-*-*` in this case.

rorth commented:
> > Besides, if you can tell me **exactly** what to check, I could do this for both Solaris and Linux. I think that would be better than blindly `XFAIL`ing the tests without understanding the issues.
> 
> The two tests are executable Fortran tests. Output for both tests lists expected output on the left side of each line, and actual output on the right. Is that output correct?

If that's the case, shouldn't the tests check if they produce this expected output?

I get that output only on Linux/x86_64: I've just compiled the test with `flang -o ieee_femodes.exe ieee_femodes.f90` in all cases.
- On Linux/x86_64, it compiles, links and runs with the expected output.
- On Solaris/amd64, the test fails to link:
  ```
  Undefined			first referenced
   symbol  			    in file
  ieee_get_modes_0_                   /var/tmp/ieee_femodes-440ad6.o
  ieee_set_modes_0_                   /var/tmp/ieee_femodes-440ad6.o
  ```
  Comparing the `.s` files between Linux and Solaris, I find that where Linux has a call to `fegetmode`, Solaris has one to `ieee_get_modes_0_` instead: not really astonishing since Solaris lacks TS 18661-1:2014 support.  It would certainly be better if `flang` wouldn't generate calls to functions known not to exist here, though.
- On both Solaris/sparcv9 and Linux/sparc64, things are way worse, however: `flang` errors out like this:
```
LLVM ERROR: Cannot select: t137: ch = set_rounding t133, t253
  t253: i32,ch = load<(dereferenceable invariant load (s8) from @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1, align 8), sext from i8> t0, t287, undef:i64
    t287: i64 = add t284, t286
      t284: i64 = shl t283, Constant:i32<12>
        t283: i64 = add t280, t282
          t280: i64 = SPISD::Hi TargetGlobalAddress:i64<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=3]
            t279: i64 = TargetGlobalAddress<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=3]
          t282: i64 = SPISD::Lo TargetGlobalAddress:i64<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=4]
            t281: i64 = TargetGlobalAddress<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=4]
        t274: i32 = Constant<12>
      t286: i64 = SPISD::Lo TargetGlobalAddress:i64<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=5]
        t285: i64 = TargetGlobalAddress<ptr @_QQroX_QM__fortran_builtinsT__builtin_ieee_round_typeX1> 0 [TF=5]
    t1: i64 = undef
In function: _QQmain
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: tools/clang/stage2-bins/bin/flang -fc1 -triple sparcv9-sun-solaris2.11 -emit-obj -fcolor-diagnostics -mrelocation-model static -resource-dir tools/clang/stage2-bins/lib/clang/20 -mframe-pointer=all -o /var/tmp/ieee_femodes-b9df56.o -x f95-cpp-input /vol/llvm/src/llvm-project/local/flang/test/Lower/Intrinsics/ieee_femodes.f90
1.	Running pass 'Function Pass Manager' on module 'FIRModule'.
2.	Running pass 'SPARC DAG->DAG Pattern Instruction Selection' on function '@_QQmain'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  flang-20  0x00000001084fe65c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 36
1  flang-20  0x00000001084fef60 SignalHandler(int) + 896
2  libc.so.1 0xffffffff7eec6068 __sighndlr + 12
3  libc.so.1 0xffffffff7eeb8910 call_user_handler + 1024
4  libc.so.1 0xffffffff7eeb8d00 sigacthandler + 208
5  libc.so.1 0xffffffff7eecb180 __lwp_sigqueue + 8
6  libc.so.1 0xffffffff7ede933c abort + 252
7  flang-20  0x000000010846502c llvm::report_fatal_error(llvm::Twine const&, bool) + 836
8  flang-20  0x000000010afa78b0 llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*) + 636
9  flang-20  0x000000010afa64ec llvm::SelectionDAGISel::SelectCodeCommon(llvm::SDNode*, unsigned char const*, unsigned int) + 26512
10 flang-20  0x0000000107d22d50 (anonymous namespace)::SparcDAGToDAGISel::Select(llvm::SDNode*) + 5780
11 flang-20  0x000000010af98708 llvm::SelectionDAGISel::DoInstructionSelection() + 1756
12 flang-20  0x000000010af974b8 llvm::SelectionDAGISel::CodeGenAndEmitDAG() + 6792
13 flang-20  0x000000010af952a4 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) + 9176
14 flang-20  0x000000010af91980 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) + 932
15 flang-20  0x000000010af8ece8 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) + 216
16 flang-20  0x000000010c028f14 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 736
17 flang-20  0x000000010dd0c05c llvm::FPPassManager::runOnFunction(llvm::Function&) + 600
18 flang-20  0x000000010dd13d78 llvm::FPPassManager::runOnModule(llvm::Module&) + 68
19 flang-20  0x000000010dd0c800 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 1192
20 flang-20  0x00000001087cc9ec Fortran::frontend::CodeGenAction::executeAction() + 2928
21 flang-20  0x00000001085584f0 Fortran::frontend::FrontendAction::execute() + 12
22 flang-20  0x0000000108543670 Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) + 356
23 flang-20  0x000000010855d114 Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) + 5192
24 flang-20  0x0000000106762b6c fc1_main(llvm::ArrayRef<char const*>, char const*) + 784
25 flang-20  0x00000001067619d0 main + 5344
26 flang-20  0x0000000106760204 _start + 100
flang-20: error: unable to execute command: Abort (core dumped)
flang-20: error: flang frontend command failed due to signal (use -v to see invocation)
flang version 20.0.0git
Target: sparcv9-sun-solaris2.11
Thread model: posix
InstalledDir: tools/clang/stage2-bins/bin
Build config: +assertions
```
  No idea if this is a SPARC backend bug or a bug in Flang.

vdonaldson commented:
Thank you @rorth for taking the time to build these compilers and provide detailed analysis of what happens in these environments.

These correctness problems exist without the code in this PR. I do not understand the details necessary to fully implement this functionality in environments that do not have access to the libm calls available in linux x86 environments. And I can't currently justify additional research or effort to that end. This iteration of the PR therefore reverts to generating a compile time "not yet implemented" message for calls to these procedures in environments that do not have this libm support. The code still addresses eventual cross compilation and floating point environment data size issues. It should provide a base for eventual implementation of this functionality in additional environments whenever someone chooses to do that. As it is not obvious to me how to exercise the two lit tests only in the appropriate cases I have deleted the tests.  I don't think these tests are very useful. There are end-to-end tests for these calls in Fortran test suites. That should allow this code to build in all environments and either generate correct executable code or a compile time message about missing functionality.

rorth commented:
Thanks.  This way, I again get clean test results on all four targets tested.

I tried reviving the last versions of `ieee_femodes.f90` and `ieee_festatus.f90`, adding explicit `-target x86_64-unknown-linux-gnu` args.  This way, they **do** `PASS` on both `x86_64-pc-linux-gnu` and `sparc64-unknown-linux-gnu`, but `FAIL`
on `amd64-pc-solaris2.11` and `sparcv9-sun-solaris2.11` with `IntrinsicCall.cpp:4507: not yet implemented: intrinsic module procedure: ieee_get_status`.  This is sort of unexpected: when targeting Linux, the tests should `PASS` on non-Linux hosts.  I believe the code in `IntrinsicCall.cpp:IntrinsicLibrary::genIeeeGetOrSetModesOrStatus` might be confused.  Right now, by checking `__GLIBC_USE_IEC_60559_BFP_EXT` it checks a **host** property, while the existance of `fe[gs]etmode/fe[gs]etenv` is a property of the **target** system.  So there may well lurk a fundamental issue here.

But as you say, it's unclear what equivalents to those TS 18661-1:2014 functions one could use (if host compiler needs to actually call them, not just the runtime), and I fully understand that you cannot justify investigating other systems.

vdonaldson commented:
Yes, I agree on all counts. The code takes steps toward eventually supporting cross compilation, but the temporary TODO in `genIeeeGetOrSetModesOrStatus` assumes host = target. The current goal for the PR is to conservatively leave the the code in a state that is correct in non cross compilation environments. The TODO may be triggered for some cases that could be handled now. This should provide an improved base for future code to build on. I'm still considering a few tweaks to the code.

vdonaldson committed:
[flang] Modifications to ieee floating point environment procedures

Intrinsic module procedures ieee_get_modes, ieee_set_modes, ieee_get_status,
and ieee_set_status store and retrieve opaque data values whose size varies
by machine and OS environment. These data values are usually, but not always
small. Their sizes are not directly known in a cross compilation environment.
Address this by implementing two mechanisms for processing these data values.
Environments that use typical small data sizes can access storage defined
at compile time. When this is not valid, data storage of any size can be
allocated at runtime. Calls in environments that do not have access to the
underlying libm functions will generate a compile time "not yet implemented"
message.

brooksdavis commented on flang/runtime/exceptions.cpp:
```diff
@@ -46,7 +42,11 @@ uint32_t RTNAME(MapException)(uint32_t excepts) {
   static constexpr uint32_t v{FE_INVALID};
-  static constexpr uint32_t s{__FE_DENORM}; // subnormal
+#if __x86_64__
+  static constexpr uint32_t s{__FE_DENORM}; // nonstandard, not a #define
```
If `__FE_DENORM` isn't a macro and isn't standard, how can we rely on it being defined? This breaks the build on FreeBSD with the base system llvm-18 based clang compiler.
